sqlair.go

```diff
@@ -0,0 +1,230 @@
+func Prepare(query string, typeInstantiations ...any) (*Statement, error) {
+	parser := expr.NewParser()
+	parsedExpr, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	preparedExpr, err := parsedExpr.Prepare(typeInstantiations...)
+	if err != nil {
+		return nil, err
+	}
+	return &Statement{pe: preparedExpr}, nil
+}
+
+// MustPrepare is the same as prepare except that it panics on error.
+func MustPrepare(query string, typeInstantiations ...any) *Statement {
+	s, err := Prepare(query, typeInstantiations...)
+	if err != nil {
+		panic(err)
+	}
+	return s
+}
+
+type DB struct {
+	db *sql.DB
+}
+
+func NewDB(db *sql.DB) *DB {
+	return &DB{db: db}
+}
+
+// sqlDB returns the underlying database object.
+func (db *DB) sqlDB() *sql.DB {
+	return db.db
+}
+
+// Exec executes an SQLair Statement without returning any rows.
+// Exec uses ExecContext with context.Background internally.
+func (db *DB) Exec(s *Statement, inputStructs ...any) (sql.Result, error) {
+	return db.ExecContext(s, context.Background())
+}
+
+// ExecContext executes an SQLair Statement without returning any rows.
+func (db *DB) ExecContext(s *Statement, ctx context.Context, inputStructs ...any) (sql.Result, error) {
+	ce, err := s.pe.Complete(inputStructs...)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := db.db.ExecContext(ctx, ce.CompletedSQL(), ce.CompletedArgs()...)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Query takes a prepared SQLair Statement and returns a Query object for
+// iterating over the results.
+// Query uses QueryContext with context.Background internally.
+func (db *DB) Query(s *Statement, inputStructs ...any) (*Query, error) {
+	return db.QueryContext(s, context.Background(), inputStructs...)
+}
+
+// QueryContext takes a prepared SQLair Statement and returns a Query object for
+// iterating over the results.
+func (db *DB) QueryContext(s *Statement, ctx context.Context, inputStructs ...any) (*Query, error) {
+	ce, err := s.pe.Complete(inputStructs...)
+	if err != nil {
+		return nil, err
+	}
+
+	q := func() (*sql.Rows, error) {
+		return db.db.QueryContext(ctx, ce.CompletedSQL(), ce.CompletedArgs()...)
+	}
+	return &Query{outputs: ce.OutputFields(), q: q}, nil
+}
+
+// Query holds a database query and is used to iterate over the results.
+type Query struct {
+	err     error
+	rows    *sql.Rows
+	outputs expr.OutputFields
+	q       func() (*sql.Rows, error)
+}
+
+// One runs a query and decodes the first row into outputStructs.
+func (q *Query) One(outputStructs ...any) error {
+	if !q.Next() {
+		return fmt.Errorf("cannot return one row: no results")
+	}
+	q.Decode(outputStructs...)
+	return q.Close()
+}
+
+// All iterates over the query and decodes all the rows.
+// It fabricates all struct instantiations needed.
+// It returns a list of rows, each row is a list of pointers to structs.
+// The structs appear in the syntactic order they appear in the query.
+func (q *Query) All() ([][]any, error) {
```

Comment:
As we discussed, we proably want to pass values into here similar to One and Decode, but let's please reduce a bit the size of the PR for now so we have less to figure out and can merge the basics before we figure the entire API out.